src/mscorlib/src/System/Text/StringBuilder.cs

```diff
@@ -999,6 +999,130 @@ void ISerializable.GetObjectData(SerializationInfo info, StreamingContext contex
             return this;
         }
 
+        public StringBuilder Append(StringBuilder value)
+        {
+            Contract.Ensures(Contract.Result<StringBuilder>() != null);
+
+            if (value != null && value.Length > 0)
+            {
+                if (value != this)
```

Comment:
To protect `ToString()` you should call some thing like this `lock(sb2) { sb1.Append(sb2.ToString()); }` or like this `lock(sb2) { sb1.Append(sb2); }` in new API. So what is difference?
New AppendCore makes bounds checking also before copy buffers. And it can throws Overflow exception like ToString does. Append has very similar code with ToString but copy sb2 not to new string buffer but to sb1 buffers. That all difference.